xarray/tests/test_backends.py

```diff
@@ -1389,6 +1389,10 @@ def test_chunk_encoding_with_dask(self):
         ds_chunk_irreg = ds.chunk({'x': (5, 5, 2)})
         with self.roundtrip(ds_chunk_irreg) as actual:
             assert (5,) == actual['var1'].encoding['chunks']
+        # re-save Zarr arrays
+        with self.roundtrip(ds_chunk_irreg) as original:
+            with self.roundtrip(original) as actual:
+                assert_identical(original, actual)
```

Comment:
Does this test cover #2300?

Reply:
I don't think it implements the NotImplementedError scenario, but the bug for my situation was that opening a dataset saved with chunks using  `to_zarr` and saving it unchanged with `to_zarr` would raise this error. The issue was that `enc_chunks_tuple` represents each chunk with an integer, while `var_chunks` represents it with a tuple. So this was the error raised in #2300 for example:

> NotImplementedError: Specified zarr chunks (63170,) would overlap multiple dask chunks ((63170, 63170, 63170, 63170, 63170, 63170, 63170, 63169),). This is not implemented in xarray yet.  Consider rechunking the data using `chunk()` or specifying different chunks in encoding.

Rechunking it, as suggested, would yield the same error. Changing the logic to allow the last element of the tuple to be less than the `zchunk` integer avoids this scenario. The test is just checking that a second roundtrip does not raise an error; it fails without this patch.

Reply:
Makes sense. Zarr's data model only allows a single integer value chunk size along each axis, while dask can have chunks of different length. 

Here is what the [zarr spec](https://zarr.readthedocs.io/en/stable/spec/v2.html#chunks) says about this topic:
>  Note that all chunks in an array have the same shape. If the length of any array dimension is not exactly divisible by the length of the corresponding chunk dimension then some chunks will overhang the edge of the array. The contents of any chunk region falling outside the array are undefined.

Reply:
That makes sense and padding the array could be another solution -- but then why does saving the dataset with irregular arrays work the first time? I haven't poked around in Zarr's internals so I assume the difference lies in the chunks argument to  `zarr_group.create()` ?

Reply:
I am ok with this.